src/java/org/apache/cassandra/service/StorageProxy.java

```diff
@@ -1248,20 +1249,22 @@ public static void mutateWithTriggers(List<? extends IMutation> mutations,
         if (augmented != null || mutateAtomically || updatesView)
             mutateAtomically(augmented != null ? augmented : (List<Mutation>)mutations, consistencyLevel, updatesView, requestTime);
         else
-            dispatchMutationsWithRetryOnDifferentSystem(mutations, consistencyLevel, requestTime);
+            dispatchMutationsWithRetryOnDifferentSystem(mutations, consistencyLevel, requestTime, preserveTimestamps);
     }
 
-    public static void dispatchMutationsWithRetryOnDifferentSystem(List<? extends IMutation> mutations, ConsistencyLevel consistencyLevel, Dispatcher.RequestTime requestTime)
+    public static void dispatchMutationsWithRetryOnDifferentSystem(List<? extends IMutation> mutations, ConsistencyLevel consistencyLevel, Dispatcher.RequestTime requestTime, boolean preserveTimestamps)
     {
         while (true)
         {
             ClusterMetadata cm = ClusterMetadata.current();
             try
             {
-                SplitMutations splitMutations = splitMutationsIntoAccordAndNormal(cm, (List<IMutation>)mutations);
+                SplitMutations<?> splitMutations = splitMutationsIntoAccordAndNormal(cm, (List<IMutation>)mutations);
                 List<? extends IMutation> accordMutations = splitMutations.accordMutations();
-                IAccordResult<TxnResult> accordResult = accordMutations != null ? mutateWithAccordAsync(cm, accordMutations, consistencyLevel, requestTime) : null;
                 List<? extends IMutation> normalMutations = splitMutations.normalMutations();
+                if (!preserveTimestamps && normalMutations != null)
```

Comment:
Add a comment explaining that if there was ever any attempt to apply part of the mutation using the eventually consistent path then we need to continue to use the timestamp used by the eventually consistent path to not end up with multiple timestamps, but if it only ever used the transactional path then we can use the transactional timestamp to get linearizability